hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,993 @@
+                             TCA src, TCA dest, TCA jmp_dest,
+                             size_t& src_count, size_t& dest_count,
+                             InstrSet& far, InstrSet& rewrites) {
+
+  Instruction* instr = Instruction::Cast(src);
+  TCA target;
+
+  /*
+   * A smashable jcc looks exactly like a smashable jmp with a preceding
+   * conditional branch. The checks below determine if there is a smashable
+   * jmp that can be optimized. It's important to not simply analyze this
+   * smashable jmp without first considering if it's part of a jcc.
+   */
+  if ((target = smashableJccTarget(src - kInstructionSize))) {
+    if (meta.smashableLocations.count(src - kInstructionSize)) {
+      target = nullptr;
+    }
+  } else if ((target = smashableJmpTarget(src))) {
+    if (meta.smashableLocations.count(src)) {
+      target = nullptr;
+    }
+  }
+
+  assertx(((uint64_t)target & 3) == 0);
+
+  if (target) {
+    TCA adjusted = rel.adjustedAddressAfter((TCA)target);
+    if (!adjusted) { adjusted = (TCA)target; }
+    int imm = (adjusted - dest) >> vixl::kInstructionSizeLog2;
+    if (is_int26(imm) && !far.count(instr)) {
+      vixl::MacroAssembler a { dest_block };
+      dest_block.setFrontier(dest);
+      a.b(imm);
+      src_count = smashableJmpLen() >> vixl::kInstructionSizeLog2; dest_count = 1;
+      for (auto i = Instruction::Cast(src);
+           i < Instruction::Cast(src + src_count * kInstructionSize);
+           i = i->NextInstruction()) {
+        rewrites.insert(i);
+      }
+      FTRACE(3,
+             "Relocated smashable at src 0x{:08x} with target 0x{:08x} to 0x{:08x} (0x{:08x})\n",
+             (uint64_t)src, (uint64_t)target, *((uint32_t*)dest), (uint64_t)dest);
+
+      return true;
+    }
+  }
+
+  /*
+   * Smashable calls are not yet supported, because we don't have any
+   * reloc debug info logged when smashing a call.
+   */
+#if 0
+  if (target) {
+    target = nullptr;
+  } else if ((target = smashableCallTarget(src))) {
+    if (meta.smashableLocations.count(src)) {
+      target = nullptr;
+    }
+  }
+
+  if (target) {
+    TCA adjusted = rel.adjustedAddressAfter((TCA)target);
+    if (!adjusted) { adjusted = (TCA)target; }
+    int        imm = (adjusted - dest) >> vixl::kInstructionSizeLog2;
+    if (is_int26(imm) && !far.count(instr)) {
+      vixl::MacroAssembler a { dest_block };
+      dest_block.setFrontier(dest);
+      a.bl(imm);
+      src_count = smashableCallLen() >> vixl::kInstructionSizeLog2; dest_count = 1;
+      for (auto i = Instruction::Cast(src);
+           i < Instruction::Cast(src + src_count * kInstructionSize);
+           i = i->NextInstruction()) {
+        rewrites.insert(i);
+      }
+      return true;
+    }
+  }
+#endif
+
+  return false;
+}
+
+/*
+ * Returns true if the source is a PC Relative instruction.
+ *   ADR/ADRP
+ *   LDR (literal)
+ *   B.<cc> (immediate) - Not yet supported as it's not used.
+ *   B (immediate)
```

Comment:
It also seems to do a bunch of other stuff (like potentially overwriting an instruction in `dest_block`) and it takes a bunch of "out" arguments by reference.  Can you document the behavior more thoroughly?